File: src/compiler/lexer.ts
// Hand-rolled Lexer and Parser for Tenuto 3.0

export enum TokenType {
  Identifier = 'Identifier',
  String = 'String',
  Number = 'Number',
  Keyword = 'Keyword',
  Symbol = 'Symbol',
  EOF = 'EOF',
}

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export class Lexer {
  private pos = 0;
  private line = 1;
  private col = 1;

  constructor(private input: string) {}

  private advance(): string {
    const char = this.input[this.pos++];
    if (char === '\n') {
      this.line++;
      this.col = 1;
    } else {
      this.col++;
    }
    return char;
  }

  private peek(): string {
    return this.input[this.pos] || '';
  }

  private skipWhitespace() {
    while (this.pos < this.input.length) {
      const char = this.peek();
      if (/\s/.test(char)) {
        this.advance();
      } else if (char === '/' && this.input[this.pos + 1] === '/') {
        while (this.pos < this.input.length && this.peek() !== '\n') {
          this.advance();
        }
      } else if (char === '%' && this.input[this.pos + 1] === '%') {
        while (this.pos < this.input.length && this.peek() !== '\n') {
          this.advance();
        }
      } else {
        break;
      }
    }
  }

  public nextToken(): Token {
    this.skipWhitespace();

    if (this.pos >= this.input.length) {
      return { type: TokenType.EOF, value: '', line: this.line, column: this.col };
    }

    const startLine = this.line;
    const startCol = this.col;
    const char = this.peek();

    // Symbols
    if (/[{}@=:<\[\]|,\(\)*]/.test(char)) {
      if (char === '|') {
        if (this.peek() === ':') {
          this.advance();
          return { type: TokenType.Symbol, value: '|:', line: startLine, column: startCol };
        }
      } else if (char === ':') {
        if (this.peek() === '|') {
          this.advance();
          return { type: TokenType.Symbol, value: ':|', line: startLine, column: startCol };
        }
      } else if (char === '[') {
        if (/[12]/.test(this.peek()) && this.input[this.pos + 1] === '.') {
          const num = this.advance();
          this.advance(); // '.'
          return { type: TokenType.Symbol, value: `[${num}.`, line: startLine, column: startCol };
        }
      }
      return { type: TokenType.Symbol, value: this.advance(), line: startLine, column: startCol };
    }

    // Strings
    if (char === '"') {
      this.advance(); // skip "
      let value = '';
      while (this.pos < this.input.length && this.peek() !== '"') {
        value += this.advance();
      }
      if (this.peek() === '"') this.advance();
      return { type: TokenType.String, value, line: startLine, column: startCol };
    }

    // Numbers (including fractions like 4/4, and durations like 4. or 4.marc, and units like ms or ticks)
    if (/[0-9]/.test(char)) {
      let value = '';
      while (this.pos < this.input.length && /[0-9./]/.test(this.peek())) {
        value += this.advance();
      }
      // Absorb units like 'ms', 's', 'ticks', '%'
      if (this.peek() === 'm' && this.input[this.pos + 1] === 's') {
        value += this.advance();
        value += this.advance();
      } else if (this.peek() === 's' && !/[a-zA-Z]/.test(this.input[this.pos + 1] || '')) {
        value += this.advance();
      } else if (this.peek() === 't' && this.input.slice(this.pos, this.pos + 5) === 'ticks') {
        for (let i = 0; i < 5; i++) value += this.advance();
      } else if (this.peek() === '%') {
        value += this.advance();
      }
      return { type: TokenType.Number, value, line: startLine, column: startCol };
    }

    // Identifiers and Keywords
    if (/[a-zA-Z_$\+\-#^]/.test(char)) {
      let value = '';
      while (this.pos < this.input.length && /[a-zA-Z0-9_$\+\-#^.]/.test(this.peek())) {
        value += this.advance();
      }
      const keywords = ['tenuto', 'meta', 'def', 'measure', 'group', 'import', 'repeat', 'var'];
      if (keywords.includes(value)) {
        return { type: TokenType.Keyword, value, line: startLine, column: startCol };
      }
      return { type: TokenType.Identifier, value, line: startLine, column: startCol };
    }

    // Fallback
    return { type: TokenType.Symbol, value: this.advance(), line: startLine, column: startCol };
  }

  public tokenize(): Token[] {
    const tokens: Token[] = [];
    let token = this.nextToken();
    while (token.type !== TokenType.EOF) {
      tokens.push(token);
      token = this.nextToken();
    }
    tokens.push(token);
    return tokens;
  }
}

File: tenutoc/src/lexer.rs
use logos::{Logos, Lexer};

/// The Token enum represents the lexical tokens of the Tenuto language.
/// 
/// F-002 Bug Fix: The lexer implements strict dot isolation for attributes.
/// For example, `4.stacc` is parsed as `Number("4")`, `Symbol(".")`, `Identifier("stacc")`
/// rather than incorrectly consuming the dot as part of a floating-point number.
#[derive(Logos, Hash, Eq, PartialEq, Clone, Debug)]
#[regex(r"\s+", logos::skip, priority = 2)]
#[regex(r"//[^\n]*", logos::skip, priority = 2)]
#[regex(r"%%[^\n]*", logos::skip, priority = 2)]
pub enum Token {
    #[token("tenuto", |lex| lex.slice().to_string())]
    #[token("meta", |lex| lex.slice().to_string())]
    #[token("def", |lex| lex.slice().to_string())]
    #[token("measure", |lex| lex.slice().to_string())]
    #[token("group", |lex| lex.slice().to_string())]
    #[token("import", |lex| lex.slice().to_string())]
    #[token("repeat", |lex| lex.slice().to_string())]
    #[token("var", |lex| lex.slice().to_string())]
    Keyword(String),

    #[token("<[")]
    VoiceOpen,

    #[token("]>")]
    VoiceClose,

    #[token("@{")]
    MapOpen,

    #[regex(r"[0-9]+(\.[0-9]+)?(ms|s|ticks)", |lex| lex.slice().to_string(), priority = 2)]
    TimeVal(String),

    #[regex(r"[a-zA-Z_$\+\-#^][a-zA-Z0-9_$\+\-#^]*", |lex| lex.slice().to_string(), priority = 2)]
    Identifier(String),

    #[regex(r"[0-9]+", lex_number, priority = 2)]
    Number(String),

    #[regex(r#""[^"]*""#, |lex| {
        let s = lex.slice();
        s[1..s.len()-1].to_string()
    })]
    String(String),

    #[token("|:", |lex| lex.slice().to_string())]
    #[token(":|", |lex| lex.slice().to_string())]
    #[token("[1.", |lex| lex.slice().to_string())]
    #[token("[2.", |lex| lex.slice().to_string())]
    #[regex(r"[^\s\w]", |lex| lex.slice().to_string(), priority = 1)]
    Symbol(String),
}

fn lex_number(lex: &mut Lexer<Token>) -> String {
    loop {
        let remainder = lex.remainder();
        let mut chars = remainder.chars();
        let Some(c) = chars.next() else { break };

        if c.is_ascii_digit() || c == '/' {
            lex.bump(c.len_utf8());
        } else if c == '.' {
            // F-002 Fix: check if it's an accessor
            let next_c = chars.next();
            let is_accessor = match next_c {
                Some(nc) => matches!(nc, 'a'..='z' | 'A'..='Z' | '_' | '$' | '+' | '-' | '#' | '^'),
                None => false,
            };
            if is_accessor {
                break; // Do not consume the dot
            } else {
                lex.bump(1);
            }
        } else {
            break;
        }
    }

    // Units
    let remainder = lex.remainder();
    if remainder.starts_with("ms") {
        lex.bump(2);
    } else if remainder.starts_with("s") {
        let after_s = &remainder[1..];
        if !after_s.starts_with(|c: char| c.is_ascii_alphabetic()) {
            lex.bump(1);
        }
    } else if remainder.starts_with("ticks") {
        lex.bump(5);
    } else if remainder.starts_with('%') {
        lex.bump(1);
    }

    lex.slice().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use logos::Logos;
    use proptest::prelude::*;

    #[test]
    fn test_strict_token_isolation() {
        // This tests the F-002 Bug Fix. 
        // 4.stacc.ff MUST be parsed as Number(4), Symbol(.), Identifier(stacc), Symbol(.), Identifier(ff)
        let source = "rh: [c#5 eb5]:4.stacc.ff |";
        let lexer = Token::lexer(source);

        let expected = vec![
            Token::Identifier("rh".to_string()),
            Token::Symbol(":".to_string()),
            Token::Symbol("[".to_string()),
            Token::Identifier("c#5".to_string()), 
            Token::Identifier("eb5".to_string()),
            Token::Symbol("]".to_string()),
            Token::Symbol(":".to_string()),
            Token::Number("4".to_string()),
            Token::Symbol(".".to_string()),
            Token::Identifier("stacc".to_string()),
            Token::Symbol(".".to_string()),
            Token::Identifier("ff".to_string()),
            Token::Symbol("|".to_string()),
        ];

        let actual: Vec<Token> = lexer.filter_map(Result::ok).collect();
        assert_eq!(actual, expected, "Lexer failed to isolate dots from identifiers/numbers.");
    }

    #[test]
    fn test_timeval_and_sigils() {
        let source = "<[ v1: c4:4.pull(15ms) ]> @{";
        let lexer = Token::lexer(source);

        let expected = vec![
            Token::VoiceOpen,
            Token::Identifier("v1".to_string()),
            Token::Symbol(":".to_string()),
            Token::Identifier("c4".to_string()),
            Token::Symbol(":".to_string()),
            Token::Number("4".to_string()),
            Token::Symbol(".".to_string()),
            Token::Identifier("pull".to_string()),
            Token::Symbol("(".to_string()),
            Token::TimeVal("15ms".to_string()),
            Token::Symbol(")".to_string()),
            Token::VoiceClose,
            Token::MapOpen,
        ];

        let actual: Vec<Token> = lexer.filter_map(Result::ok).collect();
        assert_eq!(actual, expected, "Lexer failed to parse TimeVal or compound sigils.");
    }

    // Property-based testing: Throw random garbage at the lexer to ensure it NEVER panics.
    proptest! {
        #[test]
        fn lexer_never_panics_on_random_input(s in "\\PC*") {
            let mut lexer = Token::lexer(&s);
            while let Some(_) = lexer.next() {
                // We don't care if it's Ok or Err, we just care that it doesn't panic.
            }
        }
    }
}
